examples/qtmegachatapi/megachatapplication.cpp

```diff
@@ -0,0 +1,387 @@
+{
+    QEvent* event = new GcmEvent(msg);
+    QApplication::postEvent(appDelegate, event);
+}
+
+
+MegaChatApplication::MegaChatApplication(std::string &dir)
+{
+         //Create Dir
+         //appDir = karere::createAppDir();
+
+         appDir=dir;
+
+         //Create Main Window
+         mainWin = new MainWindow();
+
+         //Create MegaAPI
+         megaApi=new ::mega::MegaApi("karere-native", appDir.c_str(), "Karere Native");
+         megaApi->setLogLevel(MegaApi::LOG_LEVEL_DEBUG);
+         megaApi->addListener(this);
+
+         //Create MegaChatAPI
+         megaChatApi=new ::megachat::MegaChatApi(megaApi);
+         megaChatApi->setLogLevel(MegaChatApi::LOG_LEVEL_DEBUG);
+         megaChatApi->addChatRequestListener(this);
+         megaChatApi->addChatListener(this);
+
+         #ifndef KARERE_DISABLE_WEBRTC
+            megaChatApi->addChatCallListener(mainWin);
+         #endif
+
+         mainWin->setMegaChatApi(megaChatApi);
+
+         //Create Login Dialog
+         loginDlg = new LoginDialog(nullptr);
+
+         //Create websocket
+         websocketsIO = new LibwsIO();
+}
+
+MegaChatApplication::~MegaChatApplication()
+{
+    megaApi->logout(nullptr);
+    megaChatApi->logout(nullptr);
+
+    marshallCall([this]()
+    {
+    delete megaChatApi;
+    delete megaApi;
+    qApp->quit();
+    delete websocketsIO;
+    delete mainWin;
+    loginDlg->destroy();
+    }, NULL);
+}
+
+int MegaChatApplication::init()
+{
+    return megaChatApi->init(sid);
+}
+
+void MegaChatApplication::login()
+{
+    auto pms = loginDlg->requestCredentials();
+    pms
+       .then([this](const std::pair<std::string, std::string>& cred)
+       {
+           assert(loginDlg);
+           loginDlg->setState(karere::IApp::ILoginDialog::kLoggingIn);
+           megaApi->login(cred.first.c_str(), cred.second.c_str());           
+       })
+       .fail([](const promise::Error& err)
+       {
+       });
+    QObject::connect(mainWin, SIGNAL(esidLogout()), &appDelegate, SLOT(onEsidLogout()));
+}
+
+void MegaChatApplication::readSid()
+{
+    char buf[256];
+    std::ifstream sidf(appDir+"/sid");
+    if (!sidf.fail())
+    {
+       sidf.getline(buf, 256);
+       if (!sidf.fail())
+           sid=strdup(buf);
+           //strcpy (sid,buf);
+    }
+}
+
+const char *MegaChatApplication::getSid() const
+{
+    return sid;
+}
+
+void MegaChatApplication::sigintHandler(int)
+{
+    printf("SIGINT Received\n"); //don't use the logger, as it may cause a deadlock
+    fflush(stdout);
+    marshallCall([this]{mainWin->close();}, NULL);
+}
+
+void MegaChatApplication::saveSid(const char* sdkSid)
+{
+    std::ofstream osidf(appDir+"/sid");
+    assert(sdkSid);
+    osidf << sdkSid;
+    osidf.close();
+}
+
+
+
+
+
+//------------------------------------------------------------------------------------------------------>
+//implementation for MegaChatListener
+void MegaChatApplication::onChatInitStateUpdate(megachat::MegaChatApi *api, int newState)
+{
```

Comment:
Note the `newState` you will receive at this callback is from `MegaChatApi`, so the state codes are different than the internal ones of karere. You need to replace the `karere::Client::kInitHasOfflineSession` by `MegaChatApi::INIT_ONLINE_SESSION` and so on

Reply:
Done